File: icr_strategy/icr/signals.py
from __future__ import annotations

from dataclasses import dataclass, asdict
from typing import Literal

import numpy as np
import pandas as pd

from .config import StrategyConfig
from .coinlegs_fusion import coinlegs_confluence_from_row
from .divergence import divergence_confluence
from .ict import ict_confluence
from .mtf import mtf_confluence
from .structure import (
    Compression,
    Impulse,
    detect_compression,
    find_recent_impulse,
    is_bearish_trend,
    is_bullish_trend,
    valid_pullback,
)

Direction = Literal["long", "short"]


@dataclass(frozen=True)
class Signal:
    symbol: str
    timeframe: str
    index: int
    timestamp: pd.Timestamp
    direction: Direction
    entry: float
    stop: float
    tp1: float
    tp2: float
    tp3: float
    score: int
    trend_score: int
    impulse_score: int
    pullback_score: int
    compression_score: int
    trigger_score: int
    volume_score: int
    rr_score: int
    rr_to_tp1: float
    reason: str
    impulse_start: int
    impulse_end: int
    compression_start: int
    compression_end: int
    ict_score: int = 0
    divergence_score: int = 0
    mtf_score: int = 0
    coinlegs_score: int = 0
    coinlegs_alpha_score: float = float("nan")
    coinlegs_bias: str = "missing"
    ict_details: str = ""
    divergence_kind: str = "none"
    predictive_tag: str = "neutral"
    mtf_details: str = ""
    coinlegs_details: str = ""

    def to_dict(self) -> dict:
        out = asdict(self)
        out["timestamp"] = self.timestamp.isoformat()
        return out


def _volume_confirmation(row: pd.Series, baseline_volume: float, direction: Direction) -> int:
    if not np.isfinite(baseline_volume) or baseline_volume <= 0:
        return 0
    vol_ratio = float(row.volume) / baseline_volume
    volume_z = float(row.volume_zscore) if "volume_zscore" in row and np.isfinite(float(row.volume_zscore)) else 0.0
    if vol_ratio >= 1.35 or volume_z >= 1.5:
        return 10
    if vol_ratio >= 1.15 or volume_z >= 1.0:
        return 8
    if vol_ratio >= 1.00:
        return 6
    return 0


def _trigger_score(row: pd.Series, compression: Compression, direction: Direction, cfg: StrategyConfig) -> tuple[bool, int, str]:
    close = float(row.close)
    ma7 = float(row.ma7)
    if direction == "long":
        level = max(compression.high, ma7)
        close_position_ok = float(row.close_position) >= cfg.candle_close_position_threshold
        decisive = close > level and close_position_ok
        distance = close - level
        reason = f"long reclaim above {level:.8f}"
    else:
        level = min(compression.low, ma7)
        close_position_ok = float(row.close_position) <= (1.0 - cfg.candle_close_position_threshold)
        decisive = close < level and close_position_ok
        distance = level - close
        reason = f"short breakdown below {level:.8f}"

    if not decisive:
        return False, 0, "no decisive reclaim/breakdown"
    atr = float(row.atr14)
    score = 10
    if atr > 0 and distance >= 0.20 * atr:
        score += 3
    if float(row.body_ratio) >= 0.50:
        score += 2
    if "displacement" in row and np.isfinite(float(row.displacement)) and float(row.displacement) >= 1.0:
        score += 1
    return True, min(15, score), reason


def _targets(direction: Direction, entry: float, stop: float, impulse: Impulse, cfg: StrategyConfig) -> tuple[float, float, float, float]:
    risk = abs(entry - stop)
    if risk <= 0:
        return float("nan"), float("nan"), float("nan"), float("nan")
    if direction == "long":
        tp1 = max(impulse.extreme, entry + cfg.tp1_r * risk)
        tp2 = max(entry + cfg.tp2_r * risk, entry + 0.75 * impulse.range_value)
        tp3 = max(entry + 5.0 * risk, entry + impulse.range_value)
        rr_to_tp1 = (tp1 - entry) / risk
    else:
        tp1 = min(impulse.extreme, entry - cfg.tp1_r * risk)
        tp2 = min(entry - cfg.tp2_r * risk, entry - 0.75 * impulse.range_value)
        tp3 = min(entry - 5.0 * risk, entry - impulse.range_value)
        rr_to_tp1 = (entry - tp1) / risk
    return tp1, tp2, tp3, rr_to_tp1


def _passes_optional_execution_filters(row: pd.Series, cfg: StrategyConfig) -> bool:
    if cfg.min_volume_ma20 is not None:
        if "volume_ma20" not in row or not np.isfinite(float(row.volume_ma20)) or float(row.volume_ma20) < cfg.min_volume_ma20:
            return False
    # CSVs can optionally include bid/ask or spread_bps. No spread data means the filter is skipped.
    if cfg.max_spread_bps is not None:
        spread_bps = None
        if "spread_bps" in row and np.isfinite(float(row.spread_bps)):
            spread_bps = float(row.spread_bps)
        elif "bid" in row and "ask" in row and np.isfinite(float(row.bid)) and np.isfinite(float(row.ask)) and float(row.close) > 0:
            spread_bps = (float(row.ask) - float(row.bid)) / float(row.close) * 10_000
        if spread_bps is not None and spread_bps > cfg.max_spread_bps:
            return False
    return True


def build_signal(
    df: pd.DataFrame,
    i: int,
    symbol: str,
    timeframe: str,
    direction: Direction,
    cfg: StrategyConfig,
) -> Signal | None:
    if direction == "long" and not cfg.allow_longs:
        return None
    if direction == "short" and not cfg.allow_shorts:
        return None

    row = df.iloc[i]
    coil_gate_delta = 0
    if cfg.enable_coil_gate:
        coil_value = float(row.coil_score) if "coil_score" in row and np.isfinite(float(row.coil_score)) else float("nan")
        if not np.isfinite(coil_value) or coil_value < cfg.min_coil_score:
            return None
        coil_gate_delta = int(max(0, min(cfg.coil_score_bonus_cap, round((coil_value - cfg.min_coil_score) / 5.0))))
    if not _passes_optional_execution_filters(row, cfg):
        return None

    trend_ok = is_bullish_trend(df, i, cfg) if direction == "long" else is_bearish_trend(df, i, cfg)
    if not trend_ok:
        return None
    trend_score = 20

    impulse = find_recent_impulse(df, i, direction, cfg)
    if impulse is None:
        return None

    pullback_ok, pullback_score, pullback_reason = valid_pullback(df, i, impulse, cfg)
    if not pullback_ok:
        return None

    compression = detect_compression(df, i, direction, cfg)
    if compression is None:
        return None
    if compression.end <= impulse.end:
        # Compression must resolve after the impulse, otherwise the setup is not sequential.
        return None

    if not np.isfinite(float(row.atr14)) or not np.isfinite(float(row.ma7)):
        return None
    trigger_ok, trigger_score, trigger_reason = _trigger_score(row, compression, direction, cfg)
    if not trigger_ok:
        return None

    volume_score = _volume_confirmation(row, compression.avg_volume, direction)
    if volume_score == 0:
        return None

    entry = float(row.close)
    atr_buffer = cfg.stop_atr_buffer * float(row.atr14)
    if direction == "long":
        stop = float(compression.low) - atr_buffer
        if stop >= entry:
            return None
    else:
        stop = float(compression.high) + atr_buffer
        if stop <= entry:
            return None

    tp1, tp2, tp3, rr_to_tp1 = _targets(direction, entry, stop, impulse, cfg)
    risk = abs(entry - stop)
    if risk <= 0:
        return None
    rr_available = (tp2 - entry) / risk if direction == "long" else (entry - tp2) / risk
    if not np.isfinite(rr_available) or rr_available < cfg.min_rr:
        rr_score = 0
    elif rr_available >= 4.0:
        rr_score = 5
    elif rr_available >= 3.0:
        rr_score = 4
    else:
        rr_score = 3

    ict = ict_confluence(df, i, direction, impulse, cfg)
    div = divergence_confluence(df, i, direction, cfg)
    mtf = mtf_confluence(df, i, direction, cfg)
    coinlegs = coinlegs_confluence_from_row(row, direction) if cfg.enable_coinlegs else coinlegs_confluence_from_row(pd.Series(dtype=object), direction)
    coinlegs_delta = max(-cfg.coinlegs_score_cap, min(cfg.coinlegs_score_cap, coinlegs.score_delta)) if cfg.enable_coinlegs else 0

    total_score = (
        trend_score
        + impulse.score
        + pullback_score
        + compression.score
        + trigger_score
        + volume_score
        + rr_score
        + ict.score_delta
        + div.score_delta
        + mtf.score_delta
        + coinlegs_delta
        + coil_gate_delta
    )
    total_score = int(max(0, min(100, total_score)))
    if total_score < cfg.score_threshold or rr_available < cfg.min_rr:
        return None

    reason_parts = [
        trigger_reason,
        pullback_reason,
        f"ICT({ict.score_delta}): {ict.details}",
        f"DIV({div.score_delta}): {div.kind}/{div.predictive_tag}",
        f"MTF({mtf.score_delta}): {mtf.details}",
        f"COINLEGS({coinlegs_delta}): alpha={coinlegs.alpha_score}, bias={coinlegs.bias}, {coinlegs.details}",
        f"COIL({coil_gate_delta}): score={float(row.coil_score) if 'coil_score' in row and np.isfinite(float(row.coil_score)) else 'missing'}",
        f"score={total_score}",
    ]

    return Signal(
        symbol=symbol,
        timeframe=timeframe,
        index=i,
        timestamp=pd.Timestamp(row.timestamp),
        direction=direction,
        entry=entry,
        stop=stop,
        tp1=tp1,
        tp2=tp2,
        tp3=tp3,
        score=total_score,
        trend_score=trend_score,
        impulse_score=impulse.score,
        pullback_score=pullback_score,
        compression_score=compression.score,
        trigger_score=trigger_score,
        volume_score=volume_score,
        rr_score=rr_score,
        rr_to_tp1=rr_to_tp1,
        reason="; ".join(reason_parts),
        impulse_start=impulse.start,
        impulse_end=impulse.end,
        compression_start=compression.start,
        compression_end=compression.end,
        ict_score=ict.score_delta,
        divergence_score=div.score_delta,
        mtf_score=mtf.score_delta,
        coinlegs_score=coinlegs_delta,
        coinlegs_alpha_score=coinlegs.alpha_score,
        coinlegs_bias=coinlegs.bias,
        ict_details=ict.details,
        divergence_kind=div.kind,
        predictive_tag=div.predictive_tag,
        mtf_details=mtf.details,
        coinlegs_details=coinlegs.details,
    )


def find_signal(df: pd.DataFrame, i: int, symbol: str, timeframe: str, cfg: StrategyConfig) -> Signal | None:
    long_signal = build_signal(df, i, symbol, timeframe, "long", cfg)
    short_signal = build_signal(df, i, symbol, timeframe, "short", cfg)
    if long_signal and short_signal:
        return long_signal if long_signal.score >= short_signal.score else short_signal
    return long_signal or short_signal

File: src/server/analysis/icr/signals.ts
import type { EnrichedCandle, IcrConfig, UnifiedSignal } from "../types";
import type { Direction } from "./structure";
import {
  isBullishTrend,
  isBearishTrend,
  findRecentImpulse,
  validPullback,
  detectCompression,
} from "./structure";
import type { Impulse, Compression } from "./structure";

/* ─── Private helpers ────────────────────────────────────────────────────── */

/**
 * Rate the volume confirmation on the trigger candle against a baseline
 * (typically the compression average volume).
 *
 * Returns 0 (fail), 6, 8, or 10.
 */
function volumeConfirmation(
  candle: EnrichedCandle,
  baselineVolume: number,
): number {
  if (!Number.isFinite(baselineVolume) || baselineVolume <= 0) return 0;
  const volRatio = candle.volume / baselineVolume;
  const volumeZ = candle.volumeZscore;
  if (volRatio >= 1.35 || volumeZ >= 1.5) return 10;
  if (volRatio >= 1.15 || volumeZ >= 1.0) return 8;
  if (volRatio >= 1.0) return 6;
  return 0;
}

/**
 * Check whether the trigger candle decisively reclaims (long) or breaks down
 * (short) through the compression boundary.
 */
function triggerScore(
  candle: EnrichedCandle,
  compression: Compression,
  direction: Direction,
  cfg: IcrConfig,
): { triggered: boolean; score: number } {
  const close = candle.close;
  const ma7 = candle.ma7;
  let level: number;
  let decisive: boolean;

  if (direction === "long") {
    level = Math.max(compression.high, ma7);
    const closePositionOk =
      candle.closePosition >= cfg.candleClosePositionThreshold;
    decisive = close > level && closePositionOk;
  } else {
    level = Math.min(compression.low, ma7);
    const closePositionOk =
      candle.closePosition <= 1.0 - cfg.candleClosePositionThreshold;
    decisive = close < level && closePositionOk;
  }

  if (!decisive) return { triggered: false, score: 0 };

  let score = 10;
  const distance =
    direction === "long" ? close - level : level - close;
  if (candle.atr14 > 0 && distance >= 0.2 * candle.atr14) score += 3;
  if (candle.bodyRatio >= 0.5) score += 2;
  if (candle.displacement >= 1.0) score += 1;

  return { triggered: true, score: Math.min(15, score) };
}

/**
 * Compute the take-profit level and the risk-to-reward ratio.
 *
 * Uses 3 R as the baseline TP multiplier. The TP is also bounded by the
 * impulse extreme and 75 % of the impulse range.
 */
function targets(
  direction: Direction,
  entry: number,
  stop: number,
  impulse: Impulse,
): { takeProfit: number; rr: number } {
  const risk = Math.abs(entry - stop);
  if (risk <= 0) return { takeProfit: entry, rr: 0 };

  let takeProfit: number;
  let rr: number;

  if (direction === "long") {
    takeProfit = Math.max(
      impulse.extreme,
      entry + 3.0 * risk,
      entry + 0.75 * impulse.rangeValue,
    );
    rr = (takeProfit - entry) / risk;
  } else {
    takeProfit = Math.min(
      impulse.extreme,
      entry - 3.0 * risk,
      entry - 0.75 * impulse.rangeValue,
    );
    rr = (entry - takeProfit) / risk;
  }

  return { takeProfit, rr };
}

function rrScore(rrValue: number, minRr: number): number {
  if (rrValue < minRr) return 0;
  if (rrValue >= 4.0) return 5;
  if (rrValue >= 3.0) return 4;
  return 3;
}

/* ─── Public API ─────────────────────────────────────────────────────────── */

/**
 * Build a single ICR signal at candle index `i`.
 *
 * Sequential gates (each must pass):
 * 1. Coil gate (optional)        — minCoilScore check + bonus delta
 * 2. Trend gate                  — 20 pts
 * 3. Impulse detection           — 0–20 pts
 * 4. Pullback validation         — 0–15 pts (min 10)
 * 5. Compression detection       — 0–15 pts (min 8)
 * 6. Compression after impulse   — must be sequential
 * 7. Trigger confirmation        — 0–15 pts
 * 8. Volume confirmation         — 6, 8, or 10 pts (0 = fail)
 * 9. Stop / target computation   — RR gate 3–5 pts
 * 10. Composite >= scoreThreshold
 *
 * @param coilScore - Optional pre-computed coil score for the coil gate.
 */
export function buildIcrSignal(
  candles: EnrichedCandle[],
  i: number,
  symbol: string,
  timeframe: string,
  direction: Direction,
  cfg: IcrConfig,
  coilScore?: number,
): UnifiedSignal | null {
  const candle = candles[i];
  if (!candle) return null;

  /* ── 1. Coil gate ──────────────────────────────────────────────────── */
  let coilGateDelta = 0;
  if (cfg.enableCoilGate) {
    if (coilScore === undefined || !Number.isFinite(coilScore)) return null;
    if (coilScore < cfg.minCoilScore) return null;
    coilGateDelta = Math.min(5, Math.round((coilScore - cfg.minCoilScore) / 5));
  }

  /* ── 1b. Momentum-exhaustion entry filter ──────────────────────────────
   * Do NOT chase a move that is already extended. Empirically validated on
   * 326 Tier-A alt trades (forward, runner exit): rejecting longs entered with
   * RSI >= 70 (shorts with RSI <= 30) raised win rate 20.9%->22.7%, avg R
   * +0.84->+1.07, total R +274.8->+292.7, Sharpe 3.47->3.79 — while preserving
   * the fat tail (max winner unchanged at 23.4R). 70/30 was the total-R and
   * Sharpe optimum across a threshold sweep. Market intelligence belongs at
   * ENTRY, not the exit (capping the exit destroys the edge). */
  const entryRsiMax = cfg.entryRsiMax ?? 70;
  const entryRsiMin = cfg.entryRsiMin ?? 30;
  if (Number.isFinite(candle.rsi14)) {
    if (direction === "long" && candle.rsi14 >= entryRsiMax) return null;
    if (direction === "short" && candle.rsi14 <= entryRsiMin) return null;
  }

  /* ── 2. Trend gate (20 pts) ────────────────────────────────────────── */
  const trendOk =
    direction === "long"
      ? isBullishTrend(candles, i, cfg)
      : isBearishTrend(candles, i, cfg);
  if (!trendOk) return null;
  const trendScore = 20;

  /* ── 3. Impulse detection (0–20 pts) ────────────────────────────────── */
  const impulse = findRecentImpulse(candles, i, direction, cfg);
  if (!impulse) return null;

  /* ── 4. Pullback validation (0–15 pts, min 10) ──────────────────────── */
  const pullback = validPullback(candles, i, impulse, cfg);
  if (!pullback.valid) return null;

  /* ── 5. Compression detection (0–15 pts, min 8) ─────────────────────── */
  const compression = detectCompression(candles, i, direction, cfg);
  if (!compression) return null;

  /* ── 6. Compression must resolve after the impulse ──────────────────── */
  if (compression.end <= impulse.end) return null;

  /* ── 7. Trigger confirmation (0–15 pts) ─────────────────────────────── */
  if (!Number.isFinite(candle.atr14) || !Number.isFinite(candle.ma7))
    return null;
  const trigger = triggerScore(candle, compression, direction, cfg);
  if (!trigger.triggered) return null;

  /* ── 8. Volume confirmation (6–10 pts, 0 = fail) ────────────────────── */
  const volumeScore = volumeConfirmation(candle, compression.avgVolume);
  if (volumeScore === 0) return null;

  /* ── 9. Entry, stop, targets & RR score (3–5 pts) ──────────────────── */
  const entry = candle.close;
  const atrBuffer = cfg.stopAtrBuffer * candle.atr14;

  let stop: number;
  if (direction === "long") {
    stop = compression.low - atrBuffer;
    if (stop >= entry) return null;
  } else {
    stop = compression.high + atrBuffer;
    if (stop <= entry) return null;
  }

  const { takeProfit, rr } = targets(direction, entry, stop, impulse);
  if (!Number.isFinite(rr) || rr <= 0) return null;

  const rrScoreVal = rrScore(rr, cfg.minRr);

  /* ── 10. Composite score ────────────────────────────────────────────── */
  const composite =
    trendScore +
    impulse.score +
    pullback.score +
    compression.score +
    trigger.score +
    volumeScore +
    rrScoreVal +
    coilGateDelta;

  if (composite < cfg.scoreThreshold || rr < cfg.minRr) return null;

  const tierA = cfg.tierAThreshold ?? 80;
  const tierB = cfg.tierBThreshold ?? 65;
  const tier: "A" | "B" | "C" =
    composite >= tierA ? "A" : composite >= tierB ? "B" : "C";

  const confidence = Math.min(1, composite / 100);

  const components: Record<string, number> = {
    trend: trendScore,
    impulse: impulse.score,
    pullback: pullback.score,
    compression: compression.score,
    trigger: trigger.score,
    volume: volumeScore,
    rr: rrScoreVal,
    coilBonus: coilGateDelta,
  };

  return {
    source: "icr",
    symbol,
    timeframe,
    direction,
    entry,
    stopLoss: stop,
    takeProfit,
    score: composite,
    tier,
    thesis: `ICR ${direction} on ${symbol} ${timeframe}: trend ${trendScore}, impulse ${impulse.score}, compression ${compression.score}, trigger ${trigger.score}`,
    components,
    structuralScore: composite - volumeScore - rrScoreVal - coilGateDelta,
    confidence,
    timestamp: candle.timestamp,
    metadata: {
      impulseStart: impulse.start,
      impulseEnd: impulse.end,
      compressionStart: compression.start,
      compressionEnd: compression.end,
      coilScore: coilScore ?? null,
      rr,
      // Impulse swing levels for the smart-exit fib engine. The impulse leg is
      // (origin -> extreme) in the trade direction; low/high normalized so
      // swingLow < swingHigh regardless of direction.
      impulseSwingLow: Math.min(impulse.origin, impulse.extreme),
      impulseSwingHigh: Math.max(impulse.origin, impulse.extreme),
    },
  };
}

/**
 * Scan only the most recent candle for valid ICR signals.
 *
 * Used by the analysis engine to avoid re-scanning historical candles
 * and reduce the effective N for multiple-testing correction.
 */
export function findLatestSignals(
  candles: EnrichedCandle[],
  symbol: string,
  timeframe: string,
  cfg: IcrConfig,
): UnifiedSignal[] {
  const signals: UnifiedSignal[] = [];
  const i = candles.length - 1;
  if (i < cfg.slowMa) return signals;

  const coilScore = candles[i].coilScore;
  for (const dir of ["long", "short"] as Direction[]) {
    const sig = buildIcrSignal(candles, i, symbol, timeframe, dir, cfg, coilScore);
    if (sig) signals.push(sig);
  }
  return signals;
}

/**
 * Scan every candle index for valid ICR signals.
 *
 * Starts after `cfg.slowMa` (MA99 warmup) and tests both long and short
 * directions at each position.
 */
export function findSignals(
  candles: EnrichedCandle[],
  symbol: string,
  timeframe: string,
  cfg: IcrConfig,
): UnifiedSignal[] {
  const signals: UnifiedSignal[] = [];
  const startIdx = cfg.slowMa; // ensure MA99 is warm
  for (let i = startIdx; i < candles.length; i++) {
    const coilScore = candles[i].coilScore;
    for (const dir of ["long", "short"] as Direction[]) {
      const sig = buildIcrSignal(candles, i, symbol, timeframe, dir, cfg, coilScore);
      if (sig) signals.push(sig);
    }
  }
  return signals;
}
